cleverhans/discretization/discretization_utils.py

```diff
@@ -0,0 +1,295 @@
+"""Utilty functions for discretizing image tensors in various ways.
+
+For the discretization, we may either use uniform buckets or supply our own
+custom buckets. One way to compute custom buckets is to use percentile
+information from the data distribution. The final discretized representation
+can either be a one-hot or a thermometer encoding. A thermometer encoding
+is of the form (1, 1, 1,..,1, 0, .., 0) with the transition from 1 to 0
+signifying which bucket it belongs to. To reduce the dimension, one may
+project back by convolving with a fixed random or trainable matrix.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+
+
+def flatten_last(x):
+  """Flatten the last two dimensions of a tensor into one.
+
+  Args:
+    x: Discretized input tensor of shape [-1, height, width, channels, levels]
+        to flatten.
+
+  Returns:
+    Flattened version of x, of shape [-1, height, width, channels * levels].
+  """
+  shape = x.get_shape().as_list()
+  new_shape = shape[:-1]
+  new_shape[-1] *= shape[-1]
+  new_shape[0] = tf.shape(x)[0]
+  flattened_x = tf.reshape(x, new_shape)
+  return flattened_x
+
+
+def unflatten_last(x, levels):
+  """Unflatten input tensor by separating the last two dimensions.
+
+  Args:
+    x: Discretized input tensor of shape [-1, height, width, channels * levels]
+        to unflatten.
+    levels: Number of levels the tensor has been discretized into.
+
+  Returns:
+    Unflattened version of x, of shape [-1, height, width, channels, levels].
+  """
+  shape = x.get_shape().as_list()
+  shape[-1] /= levels
+  shape[-1] = int(shape[-1])
+  shape.append(levels)
+  shape[0] = tf.shape(x)[0]
+  unflattened_x = tf.reshape(x, shape)
+  return unflattened_x
+
+
+def discretize_uniform(x, levels, thermometer=False):
```

Comment:
This doesn't seem to actually be called anywhere, do we actually need it?

Reply:
It's used if you want to actually discretize the inputs before calling a model.